### PR TITLE
Add retries to autoboot in ipxe script:

### PR DESCRIPTION
**Pull request by jacobweinstock**
## Description

This gives the `autoboot` (in the embedded script) a chance to overcome transient network issues or similiar.

## Why is this needed

Fixes: #

## Checklist:

I have:

- [ ] updated the documentation and/or roadmap (if required)
- [ ] added unit or e2e tests
- [ ] provided instructions on how to upgrade

### CODE REVIEW

**Review by chrisdoherty4 — APPROVED** [review]

**Comment by jacobweinstock** [comment]
Pending a basic test from @nshalman

**Comment by nshalman** [comment]
It's actually not behaving as expected for me. I'll see if I can figure out why.

File: binary/script/embed.ipxe [hunk 1]
```diff
@@ -25,7 +25,11 @@ echo Booting from net${idx}...
 autoboot net${idx}
 
 :autoboot
-autoboot
+set attempt:int32 0
+:retry
+iseq ${attempt} 5 && error-with-autoboot
```
Comment (nshalman) [inline binary/script/embed.ipxe:30]: should this be `iseq ${attempt} 5 && goto error-with-autoboot`

**Review by nshalman — CHANGES_REQUESTED** [review]
Didn't work when tested as written. comment with a possible fix.

File: binary/script/embed.ipxe (hunk 1 above)
Comment (jacobweinstock) [reply binary/script/embed.ipxe:30]: ah, yes! good catch. will update

Comment (jacobweinstock) [reply binary/script/embed.ipxe]: updated

**Comment by nshalman** [comment]
The more I think about this, the less I like the automated retries, but I do like the offer of a shell after a failure.
In the situation where a machine ought to be booting from disk, but is accidentally booting from the network, 5 retries could burn a lot of boot time. An offer of a shell to an operator watching after failure on the other hand might be hugely helpful for debugging while being less intrusive.

Also, I've been unable to get the retry logic to work. :(

**Comment by jacobweinstock** [comment]
> The more I think about this, the less I like the automated retries, but I do like the offer of a shell after a failure. In the situation where a machine ought to be booting from disk, but is accidentally booting from the network, 5 retries could burn a lot of boot time. An offer of a shell to an operator watching after failure on the other hand might be hugely helpful for debugging while being less intrusive.
> 
> Also, I've been unable to get the retry logic to work. :(

Agreed, i think this is part of the reason i had this on hold for a while. the trade-offs of retrying weren't sitting right with me.

**Comment by nshalman** [comment]
> Agreed, i think this is part of the reason i had this on hold for a while. the trade-offs of retrying weren't sitting right with me.

I'm glad you never merged it as-is.

**Comment by nshalman** [comment]
I did test the following which did appear to behave as I would expect:
```diff
diff --git a/binary/script/embed.ipxe b/binary/script/embed.ipxe
index b2c582b..5c85080 100644
--- a/binary/script/embed.ipxe
+++ b/binary/script/embed.ipxe
@@ -25,7 +25,9 @@ echo Booting from net${idx}...
 autoboot net${idx}

 :autoboot
-autoboot
+autoboot ||
+prompt --key 0x02 --timeout 2000 autoboot failed. Press Ctrl-B for the iPXE command line... && shell
+exit

 :boot-with-vlan
 set idx:int32 0
```

So I think I'm a :-1: on both the idea and the implementation of the retry logic, but I would be open to the brief offer of a shell after failure.

**Comment by jacobweinstock** [comment]
closing, please see conversation above.

**Pull request closed without merging**